docs/sphinx/user_guide/feature/resource.rst

```diff
@@ -0,0 +1,266 @@
+            only available when using ``Cuda`` or ``Hip`` resources. 
+          * The ``RAJA::resources`` namespace aliases the ``camp::resources`` namespace.
+
+Each resource has a set of underlying functionality that is synonymous across all resource types.  
+
+ ===================== ===============================================
+ Methods               Brief description
+ ===================== ===============================================
+ get_platform          Returns the underlying camp platform
+                       the resource is associated with.
+ get_event             Return an Event object for the resource from
+                       the last resource call.
+ allocate              Allocate data per the resource's given
+                       backend.
+ deallocate            Deallocate data per the resource's given
+                       backend.
+ memcpy                Perform a memory copy from a src location
+                       to a destination location from the
+                       resource's backend.
+ memset                Set memory value per the resourse's
+                       given backend.
+ wait_for              Enqueue a wait on the resource's stream/thread
+                       for a user passed event to occur.
+ ===================== ===============================================
+  
+Each resource type also defines specific backend information/functionality. For example, each
+Cuda resource contains a ``cudaStream_t`` value with an associated get method. See the 
+individual functionality for each resource in ``raja/tpl/camp/include/resource/``.
+
+
+------------
+Type-Erasure
+------------
+
+Resources can be declared in two formats: An erased resource, and a concrete resource. When 
+declared on the same backend, their underlying runtime functionality is the same. An 
+erased resource allows a user the ability to change the resourse backend at runtime. 
+
+Concrete Cuda resource::
+
+    RAJA::resources::Cuda my_cuda_res;
+
+Erased resource::
+
+    if (use_gpu)
+      RAJA::resources::Resource my_res{RAJA::resources::Cuda()};
+    else
+      RAJA::resources::Resource my_res{RAJA::resources::Host()};
+
+
+Memory allocation on resources::
+
+    int* a1 = my_cuda_res.allocate<int>(ARRAY_SIZE);
+    int* a2 = my_res.allocate<int>(ARRAY_SIZE);
+
+If ``use_gpu`` is ``true``, the underlying type of ``my_res`` is a Cuda resource, therefore ``a1`` 
+and ``a2`` will both be allocated on the GPU. If ``use_gpu`` is ``false``, then only ``a1`` is 
+allocated on the GPU, and ``a2`` is allocated on the host.
+
+
+------
+Forall
+------
+
+A resource is an optional argument to a ``RAJA::forall`` call. It is passed as the first argument 
+when the forall is templated on the execution policy. When the execution policy is passed by value,
+the resource is passed as the second argument. This maintains the order of conditions that can be 
```

Comment:
conditions doesn't seem quite right here, but I don't know what is

Reply:
types?

Reply:
maybe inputs?

Reply:
Agreed, inputs, done.